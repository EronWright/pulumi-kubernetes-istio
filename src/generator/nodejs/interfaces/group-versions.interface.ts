export interface GroupVersions {
  group: string;
  versions: string[];
}
