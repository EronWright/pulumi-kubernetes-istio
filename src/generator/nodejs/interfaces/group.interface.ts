import { VersionTS } from './version.interface';

export interface GroupTS {
  group: string;
  versions: VersionTS[];
}