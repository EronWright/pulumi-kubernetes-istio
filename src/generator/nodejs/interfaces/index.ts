export * from './schema.interface';
export * from './property.interface';
export * from './group.interface';
export * from './version.interface';
export * from './group-versions.interface';
export * from './templates';
