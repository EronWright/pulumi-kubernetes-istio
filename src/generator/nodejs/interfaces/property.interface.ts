export interface PropertyTS {
  property: string;
  type: string;
  // languageName: string;
  required?: boolean;
  description?: string;
}