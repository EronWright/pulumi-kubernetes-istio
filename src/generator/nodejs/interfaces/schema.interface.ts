import { Kind } from '../../../enums';

import { PropertyTS } from './property.interface';
// import { GroupVersionKind } from './group-version-kind.interface';

export interface SchemaTS {
  schema: string;
  apiVersion: string;
  description?: string;
  // gvk: GroupVersionKind;
  isEnum?: boolean;
  isKind?: boolean;
  type?: string;
  properties?: PropertyTS[];
  // kind?: Kind;
}
