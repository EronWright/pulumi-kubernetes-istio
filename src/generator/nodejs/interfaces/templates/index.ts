export * from './kind.interface';
export * from './member.interface';