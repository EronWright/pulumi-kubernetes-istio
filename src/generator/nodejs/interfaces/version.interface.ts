import { SchemaTS } from './schema.interface';

export interface VersionTS {
  version: string;
  schemas: SchemaTS[];
  inputs?: SchemaTS[];
}
