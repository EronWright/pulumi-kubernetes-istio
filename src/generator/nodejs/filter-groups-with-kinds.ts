import { GroupTS } from './interfaces';

export function filterGroupsWithKinds(groups: GroupTS[]): GroupTS[] {
  return groups.reduce((groupsWithKinds, { group, versions }) => {
    if (versions.some(({ schemas }) => schemas.some(({ isKind }) => isKind))) {
      const groupIndex =
        groupsWithKinds.push({
          group,
          versions: [],
        }) - 1;

      for (const { version, schemas } of versions) {
        if (schemas.some(({ isKind }) => isKind)) {
          const versionIndex =
            groupsWithKinds[groupIndex].versions.push({
              version,
              schemas: [],
            }) - 1;

          for (const schema of schemas) {
            if (schema.isKind) {
              groupsWithKinds[groupIndex].versions[versionIndex].schemas.push(schema);
            }
          }
        }
      }
    }

    return groupsWithKinds;
  }, []);
}
