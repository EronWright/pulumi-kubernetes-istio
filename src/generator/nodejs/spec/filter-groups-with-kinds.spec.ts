import { GroupTS } from '../interfaces';
import { filterGroupsWithKinds } from '../filter-groups-with-kinds';

describe('filterGroupsWithKinds', () => {
  it('should filter groups -> versions -> schemas -> kind', () => {
    const groups: GroupTS[] = [
      {
        group: 'networking',
        versions: [
          {
            version: 'v1alpha3',
            schemas: [
              {
                schema: 'DestinationRule',
                apiVersion: 'networking.istio.io/v1alpha3',
                description: null,
                isKind: true,
              },
              {
                schema: 'TrafficPolicy',
                apiVersion: 'networking.istio.io/v1alpha3',
                description: null,
                properties: [],
              },
            ],
          },
        ],
      },
      {
        group: 'authentication',
        versions: [
          {
            version: 'v1alpha1',
            schemas: [
              {
                schema: 'Policy',
                apiVersion: 'authentication.istio.io/v1alpha1',
                description: null,
                isKind: true,
              },
              {
                schema: 'TriggerRule',
                apiVersion: 'networking.istio.io/v1alpha3',
                description: null,
                properties: [],
              },
            ],
          },
        ],
      },
      {
        group: 'rbac',
        versions: [
          {
            version: 'v1alpha1',
            schemas: [
              {
                schema: 'WorkloadSelector',
                apiVersion: 'rbac.istio.io/v1alpha1',
                description: null,
                properties: [],
              },
            ],
          },
        ],
      },
    ];

    const groupsWithKinds = filterGroupsWithKinds(groups);
    expect(groupsWithKinds).toHaveLength(2);

    expect(groupsWithKinds[0]).toMatchObject({
      versions: [
        {
          schemas: [groups[0].versions[0].schemas[0]],
        },
      ],
    });

    expect(groupsWithKinds[1]).toMatchObject({
      versions: [
        {
          schemas: [groups[1].versions[0].schemas[0]],
        },
      ],
    });

    expect(groupsWithKinds).toMatchSnapshot();
  });
});
