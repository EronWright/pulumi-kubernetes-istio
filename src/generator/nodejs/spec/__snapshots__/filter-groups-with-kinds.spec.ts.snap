// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterGroupsWithKinds should filter groups -> versions -> schemas -> kind 1`] = `
Array [
  Object {
    "group": "networking",
    "versions": Array [
      Object {
        "schemas": Array [
          Object {
            "apiVersion": "networking.istio.io/v1alpha3",
            "description": null,
            "isKind": true,
            "schema": "DestinationRule",
          },
        ],
        "version": "v1alpha3",
      },
    ],
  },
  Object {
    "group": "authentication",
    "versions": Array [
      Object {
        "schemas": Array [
          Object {
            "apiVersion": "authentication.istio.io/v1alpha1",
            "description": null,
            "isKind": true,
            "schema": "Policy",
          },
        ],
        "version": "v1alpha1",
      },
    ],
  },
]
`;
